sdks/python/apache_beam/runners/portability/prism_runner.py

```diff
@@ -0,0 +1,210 @@
+    if zipfile.is_zipfile(url):
+      z = zipfile.ZipFile(url)
+      url = z.extract(
+          os.path.splitext(os.path.basename(url))[0], path=cache_dir)
+
+    # Make sure the binary is executable.
+    st = os.stat(url)
+    os.chmod(url, st.st_mode | stat.S_IEXEC)
+    return url
+
+  # Finds the bin or zip in the local cache, and if not, fetches it.
+  @classmethod
+  def local_bin(cls, url, cache_dir=None, ignore_cache=False):
+    # ignore_cache sets whether we should always be downloading and unzipping
+    # the file or not, to avoid staleness issues.
+    if cache_dir is None:
+      cache_dir = cls.BIN_CACHE
+    if os.path.exists(url):
+      _LOGGER.info('Using local prism binary from %s' % url)
+      return cls.maybe_unzip_and_make_executable(url, cache_dir=cache_dir)
+    else:
+      cached_bin = os.path.join(cache_dir, os.path.basename(url))
+      if os.path.exists(cached_bin) and not ignore_cache:
+        _LOGGER.info('Using cached prism binary from %s' % url)
+      else:
+        _LOGGER.info('Downloading prism binary from %s' % url)
+        if not os.path.exists(cache_dir):
+          os.makedirs(cache_dir)
+        try:
+          try:
+            url_read = FileSystems.open(url)
+          except ValueError:
+            url_read = urlopen(url)
+          with open(cached_bin + '.tmp', 'wb') as zip_write:
+            shutil.copyfileobj(url_read, zip_write, length=1 << 20)
+          os.rename(cached_bin + '.tmp', cached_bin)
+        except URLError as e:
+          raise RuntimeError(
+              'Unable to fetch remote prism binary at %s: %s' % (url, e))
+      return cls.maybe_unzip_and_make_executable(
+          cached_bin, cache_dir=cache_dir)
+
+  def construct_download_url(self, root_tag, sys, mach):
+    # Construct the prism download URL with the appropriate release tag.
+    # This maps operating systems and machine architectures to the compatible
+    # and canonical names used by the Go build targets.
+
+    # platform.system() provides compatible listings, so we need to filter out
+    # the unsupported versions.
+    opsys = sys.lower()
+    if opsys not in ['linux', 'windows', 'darwin']:
+      raise ValueError(
+          'Operating System "%s" unsupported for constructing a Prism release '
+          'binary URL.' % (opsys))
+
+    # platform.machine() will vary by system, but many names are compatible.
+    arch = mach.lower()
+    if arch in ['amd64', 'x86_64', 'x86-64', 'x64']:
+      arch = 'amd64'
+    if arch in ['arm64', 'aarch64_be', 'aarch64', 'armv8b', 'armv8l']:
+      arch = 'arm64'
+
+    if arch not in ['amd64', 'arm64']:
+      raise ValueError(
+          'Machine archictecture "%s" unsupported for constructing a Prism '
+          'release binary URL.' % (opsys))
+    return (
+        'https://github.com/apache/beam/releases/download/'
+        '%s/apache_beam-%s-prism-%s-%s.zip' %
+        (root_tag, self._version, opsys, arch))
```

Comment:
@lostluck As an example for `2.56.0` what is `root_tag`?

Reply:
https://github.com/lostluck/beam/blob/prismPython31403/sdks/python/apache_beam/runners/portability/prism_runner.py#L184

Reply:
Good question, which you've found the answer to, but otherwise:

In particular it's the tag used in the URL for the GitHub release page.  eg. the `untagged-dadf35fc62aedc91ebcb` in

https://github.com/apache/beam/releases/tag/untagged-dadf35fc62aedc91ebcb

or the v2.56.0 in https://github.com/apache/beam/releases/tag/v2.56.0.

Reply:
Note though that because that "untagged" version is a draft release, it's not possible to download binaries from it unless we wanted to do GitHub auth token shenanigans. We're more likely to not use draft Releases and only use the Pre-Release designation instead to work around that for validation.